### PR TITLE
Allow `storage_class_name` to be `""`

### LINKED ISSUE #102: Cannot create volume claim against NFS volume due to storageClassName
### Terraform Version

V0.11.1

### Affected Resource(s)
Please list the resources as a list, for example:
- kubernetes_persistent_volume_claim

### Terraform Configuration Files
```hcl
## NFS Server Service
resource "kubernetes_service" "jupyter-nfs" {
  metadata {
    name = "jupyter-nfs"
  }

  spec {
    selector {
      app = "${kubernetes_pod.jupyter-nfs.metadata.0.labels.app}"
    }

    port {
      name = "nfs"
      port = 2049
    }

    port {
      name = "mountd"
      port = 20048
    }

    port {
      name = "rpcbind"
      port = 111
    }
  }
}

## NFS Server as persistent volume
resource "kubernetes_persistent_volume" "nfs-volume" {
  metadata {
    name = "nfs-volume"
  }
  spec {
    capacity {
      storage = "5Gi"
    }
    access_modes = ["ReadWriteMany"]
    persistent_volume_source {
      nfs {
        path   = "/exports"
        server = "${kubernetes_service.jupyter-nfs.metadata.0.name}"
      }
    }
  }
}

resource "kubernetes_persistent_volume_claim" "nfs-volume" {
  metadata {
    name = "nfs-volume"
  }
  spec {
    access_modes = ["ReadWriteMany"]
    storage_class_name = ""
    resources {
      requests {
        storage = "5Gi"
      }
    }
  }
}
```

### Expected Behavior
Should be able to create a PVC the same as on 
https://github.com/kubernetes/examples/blob/master/staging/volumes/nfs/nfs-pvc.yaml

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: nfs
spec:
  accessModes:
    - ReadWriteMany
  storageClassName: ""
  resources:
    requests:
      storage: 1Mi
```

### Actual Behavior
Cannot pass empty string for storage_class_name  as noted here: http://blog.kubernetes.io/2017/03/dynamic-provisioning-and-storage-classes-kubernetes.html

Instead Storage class is set to standard (GCE) 

PVC is created an in Pending state forever, but if using `kubectl` with the above YAML is created just fine. 

### Steps to Reproduce
1. `terraform apply`

### Important Factoids
Ran on GCE using GKE

### LINKED ISSUE #339: kubernetes_persistent_volume_claim can not set storage class name
```
resource "kubernetes_persistent_volume_claim" "default" {
  metadata {
    name = "default"
  }

  spec {
    access_modes = ["ReadWriteMany"]

    resources {
      requests {
        storage = "1Gi"
      }
    }

    storage_class_name = ""
    volume_name        = "${kubernetes_persistent_volume.default.metadata.0.name}"
  }
}
```

Will always result storageclass: standard.

related https://github.com/terraform-providers/terraform-provider-kubernetes/issues/102

### LINKED ISSUE #567: PVC storage_class_name empty sets default storage class
Hello,

There is a critical bug for the kubernetes terraform provider that has been reported since 2017 but is still open.

See:
https://github.com/terraform-providers/terraform-provider-kubernetes/issues/102
https://github.com/terraform-providers/terraform-provider-kubernetes/issues/339

What happens is that terraform is not passing the empty storage class string instead somehow the pvc's are created with the default storage class. The work around mentioned by using label selectors is not working anymore on terraform 0.12.

Best regards,
David

### PR DESCRIPTION
**Pull request by lawliet89**
According to the docs at https://kubernetes.io/docs/concepts/storage/persistent-volumes/#class-1

> PVCs don’t necessarily have to request a class. A PVC with its storageClassName set equal to "" is always interpreted to be requesting a PV with no class, so it can only be bound to PVs with no class (no annotation or one set equal to ""). A PVC with no storageClassName is not quite the same and is treated differently by the cluster depending on whether the DefaultStorageClass admission plugin is turned on.

- Fixes #567
- Fixes #102
- Fixes #339

### CODE REVIEW

**Comment by lawliet89** [comment]
@appilon @alexsomesan Please take a look at this.

**Comment by saiaman** [comment]
+1 for solving this quickly ..... 
@lawliet89 looks like there's some conflicts, maybe this is the problem no?

**Comment by lawliet89** [comment]
@saiaman I have fixed the merge conflict.

IIRC this is the **third time** I have fixed conflicts on this PR. I would greatly appreciate if the maintainers could spend a bit of time to look this through.  I don't have the time nor energy to constantly check back to resolve conflicts on a PR that has been sitting around for months.

**Comment by saiaman** [comment]
Can someone please merge this fix ? it's very important

@appilon @alexsomesan

File: kubernetes/structure_persistent_volume_spec.go [hunk 1]
```diff
@@ -373,9 +373,9 @@ func flattenPersistentVolumeSpec(in v1.PersistentVolumeSpec) []interface{} {
 	if in.PersistentVolumeReclaimPolicy != "" {
 		att["persistent_volume_reclaim_policy"] = in.PersistentVolumeReclaimPolicy
 	}
-	if in.StorageClassName != "" {
-		att["storage_class_name"] = in.StorageClassName
-	}
+
+	att["storage_class_name"] = in.StorageClassName
```
Comment (pdecat) [inline kubernetes/structure_persistent_volume_spec.go]: This should not be a breaking change requiring resource re-creation as `storage_class_name` is marked `Optional` and `Computed`:

https://github.com/terraform-providers/terraform-provider-kubernetes/blob/v1.10.0/kubernetes/schema_persistent_volume_claim.go#L81-L87

**Comment by pdecat** [comment]
Maybe adding an additional test case would be a good thing?

The doc could probably also be improved to make this more explicit, currently:

> * `storage_class_name` - (Optional) Name of the storage class requested by the claim

See https://github.com/terraform-providers/terraform-provider-kubernetes/blame/v1.10.0/website/docs/r/persistent_volume_claim.html.markdown#L88

**Comment by saiaman** [comment]
@pdecat maybe you can help about it ? It's really a pain that we can't claim a filestore volume on gcloud ....

**Comment by ThinkBriK** [comment]
any chance this could be released soon ? If it's not possible, you can't bind a manually provisionned volume in kubernetes on aws.

**Comment by AlexShuraits** [comment]
@pdecat @lawliet89  This issue is really painful for filestore users and other cases I am not aware of. I have a suggestion for a quick fix by adding additional key-value, like storage_class_name = "empty" and setting the value to "". Kind of workaround, but it won't break anything.

**Comment by lawliet89** [comment]
@AlexShuraits This needs @alexsomesan who is the Hashicorp Maintainer to review and merge.

**Comment by jrhouston** [comment]
Can we add a test case for this?

**Comment by lawliet89** [comment]
Adding a bunch of tests. I will edit this comment with their test output as I go along because each test takes a while to run and costs money.

**Comment by dak1n1** [comment]
I was hoping to review this and get it merged, but I ran into something unexpected during testing. So I wrote about it on the original issue here: https://github.com/terraform-providers/terraform-provider-kubernetes/issues/102#issuecomment-616897253

Unless I'm misunderstanding something, it appears that we can already set `storage_class_name = ""`. Please let me know if I've misunderstood or tested it wrong. Thanks!

**Comment by lawliet89** [comment]
You can set `storage_class_name` to whatever you want in the configuration, including `""`.

But the provider does not actually set it to `""` in the request it sends to the Kubernetes master. It sends `nil`. And this falls back to the default storage class, which is completely different from `""`.

Consider the following configuration (please ignore the `count = 1`):

```hcl
data "google_compute_zones" "available" {
  provider = google-beta
}

resource "google_compute_region_disk" "raft" {
  provider = google-beta
  count    = 1

  name = "vault-data-${count.index}"
  type = "pd-ssd"
  size = 100

  region        = "asia-southeast1"
  replica_zones = [element(data.google_compute_zones.available.names, count.index), element(data.google_compute_zones.available.names, count.index + 1)]

  description = "Vault server data disks replica ${count.index}"
}

resource "google_compute_resource_policy" "raft_backup" {
  provider = google-beta
  count    = 1

  name   = "vault-data"
  region = "asia-southeast1"

  snapshot_schedule_policy {
    schedule {
      daily_schedule {
        days_in_cycle = 1
        start_time    = "19:00"
      }
    }

    retention_policy {
      max_retention_days    = 14
      on_source_disk_delete = "KEEP_AUTO_SNAPSHOTS"
    }

    snapshot_properties {
      storage_locations = ["asia-southeast1"]
      guest_flush       = false
    }
  }
}

resource "google_compute_region_disk_resource_policy_attachment" "raft_backup" {
  provider = google-beta
  count    = 1

  name   = google_compute_resource_policy.raft_backup[0].name
  disk   = google_compute_region_disk.raft[count.index].name
  region = google_compute_region_disk.raft[count.index].region
}

resource "kubernetes_persistent_volume" "raft" {
  count = 1

  metadata {
    name = "vault-data-${count.index}"
  }

  spec {
    access_modes = ["ReadWriteOnce"]

    capacity = {
      storage = "100G"
    }

    storage_class_name = ""

    persistent_volume_source {
      gce_persistent_disk {
        pd_name = google_compute_region_disk.raft[count.index].name
        fs_type = "ext4"
      }
    }
  }
}

resource "kubernetes_persistent_volume_claim" "raft" {
  count = 1

  metadata {
    name = "vault-data-${count.index}"
  }

  spec {
    access_modes = ["ReadWriteOnce"]
    volume_name  = kubernetes_persistent_volume.raft[count.index].metadata[0].name

    storage_class_name = ""

    resources {
      requests = {
        storage = "100G"
      }
    }
  }
}

```

with the following versions:

```
Terraform v0.12.24
+ provider.google v3.18.0
+ provider.google-beta v3.18.0
+ provider.kubernetes v1.11.1
```

The PV created is 

```yaml
apiVersion: v1
kind: PersistentVolume
metadata:
  creationTimestamp: "2020-04-21T08:18:55Z"
  finalizers:
  - kubernetes.io/pv-protection
  labels:
    app: vault
    failure-domain.beta.kubernetes.io/region: asia-southeast1
    failure-domain.beta.kubernetes.io/zone: asia-southeast1-a__asia-southeast1-b
    terraform: "true"
  name: vault-data-0
  resourceVersion: "173331496"
  selfLink: /api/v1/persistentvolumes/vault-data-0
  uid: c21d6406-2524-4af5-ac5d-aca56c6b5cb9
spec:
  accessModes:
  - ReadWriteOnce
  capacity:
    storage: 100G
  gcePersistentDisk:
    fsType: ext4
    pdName: vault-data-0
  nodeAffinity:
    required:
      nodeSelectorTerms:
      - matchExpressions:
        - key: failure-domain.beta.kubernetes.io/zone
          operator: In
          values:
          - asia-southeast1-a
          - asia-southeast1-b
        - key: failure-domain.beta.kubernetes.io/region
          operator: In
          values:
          - asia-southeast1
  persistentVolumeReclaimPolicy: Retain
  volumeMode: Filesystem
status:
  phase: Available
```

The PVC created is permanently stuck in the "pending" stage. Notice that the `storageClassName` is `standard` despite having me set it to `""`.

```yaml
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  creationTimestamp: "2020-04-21T08:33:43Z"
  finalizers:
  - kubernetes.io/pvc-protection
  labels:
    app: vault
    terraform: "true"
  name: vault-data-0
  namespace: core
  resourceVersion: "173335882"
  selfLink: /api/v1/namespaces/core/persistentvolumeclaims/vault-data-0
  uid: 699f2547-6182-40c4-8d15-1b93bd86c028
spec:
  accessModes:
  - ReadWriteOnce
  resources:
    requests:
      storage: 100G
  storageClassName: standard
  volumeMode: Filesystem
  volumeName: vault-data-0
status:
  phase: Pending
```

**Comment by dak1n1** [comment]
Thanks, that explanation was really helpful! I see what you mean. I support adding this change, to allow users to explicitly set the storage_class_name to "". However, I also would like to see the old behavior supported as well, which allows users to provision a PVC without specifying a storageclass. For example, here I've completely omitted the storageclass attribute:

```
resource "kubernetes_persistent_volume_claim" "example" {
  metadata {
    name = "mariadb-data"
    namespace = "nginx"
  }
  spec {
    access_modes = ["ReadWriteOnce"]
    resources {
      requests = {
        storage = "1Gi"
      }
    }
  }
}
```

When omitting the `storage_class_name` attribute, the default storageclass should be used. This is consistent with the behavior of kubectl:

```
[dakini@dax issue_102]$ cat pvc.yaml 
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  name: testpvc
spec:
  accessModes:
    - ReadWriteOnce
  resources:
    requests:
      storage: 1Gi


[dakini@dax issue_102]$ kubectl create -f pvc.yaml
persistentvolumeclaim/testpvc created

[dakini@dax issue_102]$ kubectl get pvc
NAME      STATUS   VOLUME                                     CAPACITY   ACCESS MODES   STORAGECLASS   AGE
testpvc   Bound    pvc-7e9303e1-81f7-4a43-9ed4-85f01e947613   1Gi        RWO            standard       6s
```

**Comment by lawliet89** [comment]
Let me figure out how to fix this. Terraform automatically turns `nil` fields into the "default" value for the specific type when expanding `schema.Resource` fields into a `map[string]interface{}`. So I need to find a "clean" way to get around this.

**Comment by lawliet89** [comment]
I've experimented further and I found nothing in the Terraform SDK that allows providers to differentiate between a `null` value vs a "zero value (`""`). I've tried `d.GetOk` and `d.GetOkExists`.

So I suggest one of the following:

- Separate "empty" PVC/PV resource
- Another boolean field that `ConflictsWith` `storage_class_name` that allows us to set an empty storage class
- A magic string that users can set to inform the provider that the user intends to use it with an empty string.

I think the second solution might be the "least bad".

**Comment by dak1n1** [comment]
Thanks for looking into possible solutions. I'll start testing these approaches and see if I can get something to work.

**Comment by lawliet89** [comment]
I just discovered the new Hashicorp Discuss forum and I posted a question there: https://discuss.hashicorp.com/t/differentiate-between-null-and-empty-value/8134

**Comment by dak1n1** [comment]
I talked this over with my team during a meeting this week, and we decided that it's best to wait until the new version of the SDK is released so we can support both the null and blank values for the storageclass. I really appreciate all the work you put into this PR. It's unfortunate that we can't move forward with it at this time. But we might be able to use parts of it in the future.

In the mean time, the work-around for this issue is to create a new storageclass and reference it in the PVC. I have an example config here, in case that helps anyone. https://github.com/terraform-providers/terraform-provider-kubernetes/pull/840

**Pull request closed without merging**